Add details to CONTRIBUTING

spbail opened the pull request:
Some fairly big changes to the contribution instructions:
- restructured the sections
- added instructions for docs-only contribution
- added step-by-step instructions for code contributions including forks etc

I tested these with a new github account and I'm pretty confident they make sense, but there's definitely a chance I missed a step. I anticipate iterations on this for both clarity and correctness, but this should be a good starting point for more detailed instructions.

Commit:
Add details to CONTRIBUTING

Some fairly big changes to the contribution instructions:
- restructured the sections
- added instructions for docs-only contribution
- added step-by-step instructions for code contributions including forks etc

I tested these with a new github account and I'm pretty confident they make sense, but there's definitely a chance I missed a step. I anticipate iterations on this for both clarity and correctness, but this should be a good starting point for more detailed instructions.

jcampbell commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,120 @@
+If you would like to run the tests without the need for local backend setups, adding the following flags will skip the respective backends:
+- `--no-postgresql` will skip postgres tests
+- `--no-spark` will skip spark tests 
+- `--no-sqlalchemy` will skip sqlalchemy tests
```
Two small clarifications here: 
1. even with all of these flags set, the *pandas* backend *will* be tested...it's not currently possible to disable those tests.
2. `no-sqlalchemy` is a superset of `no-postgresql`...if you set no-sqlalchemy then no-postgresql is implied.

jcampbell commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,120 @@
-
-Second, if you have opinions on the testing framework, we'd love to hear them! Feedback based on your perspective and experience is very welcome.
+* For now, write tests in whatever style suits your fancy. We (the core contributors) will worry about refactoring them later. As long as your thing works and is well-tested, you're good. (This is **not** an excuse to avoid writing tests. All contributions must be under test. We're just not dogmatic about the style of those tests today.)
+* Second, if you have opinions on the testing framework, we'd love to hear them! Feedback based on your perspective and experience is very welcome.
```
OBE! We should have removed this...

jcampbell commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,120 @@
+* Second, if you have opinions on the testing framework, we'd love to hear them! Feedback based on your perspective and experience is very welcome.
 
 Most of the discussion to date is encapsulated here: https://github.com/great-expectations/great_expectations/issues/167. The `refactor_tests` branch is intended as a pilot implementation.
 
```
ditto

jcampbell commented on CONTRIBUTING.md:
```diff
@@ -1,65 +1,120 @@
+Ensure any new features or behavioral differences introduced by your changes are documented in the docs, and ensure you have docstrings on your contributions. We use the Sphinx's Napoleon extension to build documentation from Google-style docstrings (see http://www.sphinx-doc.org/en/master/ext/napoleon.html).
 
 Expectations aren't just tests---they're also a kind of data documentation. Because we want expectations to be easy to interpret, we're avoiding abbreviations almost everywhere. We're not entirely consistent about this yet, but there's pretty strong consensus among early team and users that we should be heading in that direction.
 
```
remove, OBE

abegong reviewed (commented):
These changes (and James' clarifications) look great to me.

Commit:
Update CONTRIBUTING.md

Making the changes suggested by @jcampbell. We might want to add some more details to "writing tests" at a later point (or now if you have something to add!)

spbail commented:
TIL what OBE means.
